lib/src/tasks/coverage/api.dart

```diff
@@ -406,45 +417,98 @@ class CoverageTask extends Task {
-      // well, but it is sometimes malformed. In those cases, the correct
-      // observatory port is output to stdout. So we listen to both.
-      int observatoryPort;
-      process.stdout.listen((line) {
-        _coverageOutput.add('    $line');
-        if (line.contains(_observatoryPortPattern)) {
-          Match m = _observatoryPortPattern.firstMatch(line);
-          observatoryPort = int.parse(m.group(2));
-        }
-      });
-      await for (String line in process.stderr) {
-        _coverageOutput.add('    $line');
-        if (line.contains(_observatoryFailPattern)) {
-          throw new CoverageTestSuiteException(file.path);
-        }
-        if (line.contains(_observatoryPortPattern)) {
-          Match m = _observatoryPortPattern.firstMatch(line);
-          observatoryPort = int.parse(m.group(2));
-        }
-        if (line.contains(_testsFailedPattern)) {
-          throw new CoverageTestSuiteException(file.path);
+      try {
+        String testPath;
+        if (pubServe) {
+          _coverageOutput.add('Starting pub server...');
+
+          // Start `pub serve` on the `test` directory.
+          pubServeTask = startPubServe(additionalArgs: ['test']);
+
+          _coverageOutput.add('::: ${pubServeTask.command}');
+          String indentLine(String line) => '    $line';
+
+          var startupLogFinished = new Completer();
+          pubServeTask.stdOut
+              .transform(until(startupLogFinished.future))
+              .map(indentLine)
+              .listen(_coverageOutput.add);
+          pubServeTask.stdErr
+              .transform(until(startupLogFinished.future))
+              .map(indentLine)
+              .listen(_coverageErrorOutput.add);
+
+          PubServeInfo serveInfo = await pubServeTask.serveInfos.first;
+          if (!path.isWithin(serveInfo.directory, htmlFile.path)) {
+            throw '`pub serve` directory does not contain test file: ${htmlFile.path}';
+          }
+
+          var relativeHtmlPath =
+              path.relative(htmlFile.path, from: serveInfo.directory);
+          testPath = 'http://localhost:${serveInfo.port}/$relativeHtmlPath';
+
+          startupLogFinished.complete();
+          pubServeTask.stdOut.map(indentLine).join('\n').then((stdOut) {
+            if (stdOut.isNotEmpty) {
+              _coverageOutput
+                  .add('`${pubServeTask.command}` (buffered stdout)');
+              _coverageOutput.add(stdOut);
+            }
+          });
+          pubServeTask.stdErr.map(indentLine).join('\n').then((stdErr) {
+            if (stdErr.isNotEmpty) {
+              _coverageOutput
+                  .add('`${pubServeTask.command}` (buffered stdout)');
+              _coverageOutput.add(stdErr);
+            }
+          });
+        } else {
+          testPath = htmlFile.path;
         }
-        if (line.contains(_testsPassedPattern)) {
-          break;
+
+        // Run the test in content-shell.
```

Comment:
This diff is easier to look at if you [ignore whitespace changes](https://github.com/Workiva/dart_dev/pull/92/files?w=1#diff-9936a72e6f0ed192f6b4b9baa40767e0R471).